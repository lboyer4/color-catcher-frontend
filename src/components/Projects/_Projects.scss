.project-holder {
	width: 40%;
	min-height: 200px;
	height: 100%;
	background-color: #6A7062;
	overflow-y: scroll;
	color: #F9F4F5;
	text-align: center;
}

.project-name {
	border: 1px solid #012622;
	background-color: #F9F4F5;
	color: #012622;
	margin: 0;
	padding: 5px;
	height: 50px;
}

.delete-button {
	width: 25px;
	height: 20px;
	margin-left: 5px;
	background-color: #F9F4F5;
	color: #012622;
	border-radius: 2px;
	border: 1px solid #012622;
}

.delete-button:hover {
	background-color: #6A7062;
	color: #F9F4F5;
	height: 22px;
	width: 26px;
	border: 2px solid #F9F4F5;
}