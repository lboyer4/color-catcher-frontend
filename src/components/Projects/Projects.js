import React, { Component } from 'react';
import './_Projects.scss';
import { connect } from 'react-redux';
import { setProjects, deleteProject, addProject, pickProject, setPalettes } from '../../actions';
import PropTypes from 'prop-types';

export class Projects extends Component {
	constructor() {
		super();
		this.state = {
			title: '',
		}
	}

	componentDidMount = () => {
		fetch('http://localhost:3001/api/v1/project')
		.then(response => response.json())
		.then(projects => this.props.setProjects(projects))

		fetch('http://localhost:3001/api/v1/palettes')
		.then(response => response.json())
		.then(palettes => this.props.setPalettes(palettes))
	}
	
	handleChange = (e) => {
		this.setState({title: e.target.value});  
  }

	makeProject = (e) => {
		e.preventDefault();
		this.postProject(this.state.title);
		this.setState({title: ''});
	}

	postProject = (title) => {
		let name = title
		let options = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        name: name
      })
    };

		fetch('http://localhost:3001/api/v1/project', options)
		.then(response => {
			if(!response.ok) {
				throw Error('Error posting project')
			} else {
				return response
			}
		})
		.then(result => result.json())
		.then(body => this.props.addProject({id: body.id, name}))
	}

	deleteProject = (e) => {
		let options = {
      method: "DELETE"
    };
		const idToDelete = e.target.parentElement.getAttribute('data-key')
		this.props.deleteProject(idToDelete)
		fetch(`http://localhost:3001/api/v1/project/${idToDelete}`, options)
	}

	chooseProject = (e) => {
		const id = e.target.parentElement.getAttribute('data-key')
		const project = this.props.projects.find(project => {
			return project.id === Number(id)
		})
		this.props.pickProject(project)
	}

	render() {
		const projectNames = this.props.projects.length ? this.props.projects.map(project => {
			return (
				<div data-key={project.id} key={project.id}>
					<h4 onClick={this.chooseProject}>{project.name}<button className="delete-button" onClick={this.deleteProject}>
						X
					</button></h4>
					
				</div> 
			)
		}) : ''

		return (
			<div className='project-holder'>
				<h1 className='project-name'>Your Projects</h1>
				{projectNames}
				<form> 
					<input
						type='text'
	          onChange={this.handleChange}
	          name='title'
	          placeholder="Name Your Project"
	          value={this.state.title}
	         />
					<button onClick={this.makeProject}>Create</button> 
				</form>
			</div>
		)
	}
}

export const mapStateToProps = (state) => ({
	projects: state.projects,
	project: state.project
});

export const mapDispatchToProps = (dispatch) => ({
	setProjects: (projects) => dispatch(setProjects(projects)),
	addProject: (project) => dispatch(addProject(project)),
	pickProject: (project) => dispatch(pickProject(project)),
	deleteProject: (id) => dispatch(deleteProject(id)),
	setPalettes: (palettes) => dispatch(setPalettes(palettes))
});

export default connect(mapStateToProps, mapDispatchToProps)(Projects);

Projects.propTypes = {
	addProject: PropTypes.func,
	deleteProject: PropTypes.func,
	pickProject: PropTypes.func,
	project: PropTypes.object,
	projects: PropTypes.array,
	setPalettes: PropTypes.func,
	setProjects: PropTypes.func
}