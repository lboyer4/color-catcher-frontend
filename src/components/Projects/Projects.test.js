import React from 'react';
import { shallow } from 'enzyme';
import { Projects, mapStateToProps, mapDispatchToProps } from './Projects'
import { setProjects, deleteProject, addProject } from '../../actions';

describe('Projects', () => {
	let wrapper;
	let mockSetProjects;
	let mockDeleteProject;
	let mockAddProject;
	let mockProjects;

	beforeEach(() => {
		mockAddProject = jest.fn();
		mockDeleteProject = jest.fn();
		mockSetProjects = jest.fn();
		mockProjects = [{
			name: 'project one',
			id: 1
		}]

		wrapper = shallow(
			<Projects 
				setProjects={mockSetProjects}
				addProject={mockAddProject}
				deleteProject={mockDeleteProject}
				projects={mockProjects}
			/>
		);
	});

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('handleChange', () => {
		it('should set state with title', () => {

			const mockTitleEvent = {
				target: {
					value: 'Project One'
				}
			};

			const wrapper = shallow(
				<Projects 
					projects={mockProjects}
				/>
			);

			wrapper.instance().handleChange(mockTitleEvent);
			expect(wrapper.state().title).toEqual('Project One');
		});
	}); 

	describe('makeProject', () => {
		it('should call post project', () => {
			const mockMakeProjectEvent = {
				preventDefault: () => {}
			};

			wrapper.instance().postProject = jest.fn();
			wrapper.instance().makeProject(mockMakeProjectEvent);

			expect(wrapper.instance().postProject).toHaveBeenCalled();
			expect(wrapper.state().title).toEqual('');
		});
	});

	describe('deleteProject', () => {
		it.skip('should call deleteProject', () => {
			const mockTitleEvent = {
				target: {
					getAttribute: () => {
						return 1 }
				}
			};
			
			wrapper.instance().deleteProject(mockTitleEvent)
			expect(wrapper.instance().props.deleteProject).toHaveBeenCalledWith(1)
		});
	});

	describe('mapDispatchToProps', () => {
		it('should call dispatch when using setProjects', () => {

			const mockProjects = [
				{
					name: 'project one',
					id: 1
				},
				{
					name: 'project two',
					id: 2
				}
			]

			const mockDispatch = jest.fn()

			const actionToDispatch = setProjects(mockProjects)

			const mappedProps = mapDispatchToProps(mockDispatch);

			mappedProps.setProjects(mockProjects)

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)

		});

		it('should call dispatch when using addProject', () => {

			const mockProject = {
				name: 'project one',
				id: 1
			};

			const mockDispatch =jest.fn();

			const actionToDispatch = addProject(mockProject);

			const mappedProps = mapDispatchToProps(mockDispatch);

			mappedProps.addProject(mockProject);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});

		it('should call dispatch when using deleteProject', () => {

			const mockId = 1;

			const mockDispatch = jest.fn();

			const actionToDispatch = deleteProject(mockId);

			const mappedProps = mapDispatchToProps(mockDispatch);

			mappedProps.deleteProject(mockId);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});
	});
});



