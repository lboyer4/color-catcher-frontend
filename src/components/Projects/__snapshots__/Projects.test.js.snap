// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects should match the snapshot 1`] = `
<div
  className="project-holder"
>
  <h1
    className="project-name"
  >
    Your Projects
  </h1>
  <form>
    <input
      className="create-project"
      name="title"
      onChange={[Function]}
      placeholder="Name Your Project"
      type="text"
      value=""
    />
    <button
      className="make-project-btn"
      onClick={[Function]}
    >
      Create
    </button>
  </form>
  <div
    className="project"
    data-key={1}
    key="1"
  >
    <h4
      onClick={[Function]}
    >
      project one
    </h4>
    <button
      className="delete-button"
      onClick={[Function]}
    >
      X
    </button>
  </div>
</div>
`;
